### PR TITLE
[enhancement] add dlpack queue extraction and data conversions

### PR DESCRIPTION
**Pull request by icfaust**
## Description

Work in progress.

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [ ] I have reviewed my changes thoroughly before submitting this pull request.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [ ] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [ ] I have added a respective label(s) to PR if I have a permission for that.
- [ ] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [ ] I have run it locally and tested the changes extensively.
- [ ] All CI jobs are green or I have provided justification why they aren't.
- [ ] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [ ] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [ ] I have provided justification why performance has changed or why changes are not expected.
- [ ] I have provided justification why quality metrics have changed or why changes are not expected.
- [ ] I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

### CODE REVIEW

File: onedal/datatypes/_dlpack.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            raise TypeError(f"cannot move {type(obj)} to cpu")
+
+    # convert to numpy
+    if hasattr(obj, "__array__"):
+        # `copy`` param for the `asarray`` is not set.
+        # The object is copied only if needed
+        obj = np.asarray(obj)
+    else:
+        # requires numpy 1.23
+        obj = np.from_dlpack(obj)
```
Comment (Vika-F) [inline onedal/datatypes/_dlpack.py]: Should there be something like:
```
 if np.__version__ > 1.23:
     obj = np.from_dlpack(obj)
 else:
     raise SomeException
```

File: onedal/datatypes/_dlpack.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+            obj = obj.to_device("cpu")
+        elif hasattr(obj, "to"):
+            # pytorch-specific fix as it is not array api compliant
+            obj = obj.to("cpu")
+        else:
+            raise TypeError(f"cannot move {type(obj)} to cpu")
+
+    # convert to numpy
+    if hasattr(obj, "__array__"):
+        # `copy`` param for the `asarray`` is not set.
```
Comment (Vika-F) [inline onedal/datatypes/_dlpack.py]:
```suggestion
        # ``copy`` param for the ``asarray`` is not set.
```

File: onedal/utils/_sycl_queue_manager.py [hunk 3]
```diff
@@ -91,6 +97,32 @@ def fallback_to_host():
...
+        return None
+    elif device_type != backend.kDLOneAPI:
+        # Data exists on a non-SYCL, non-CPU device. This will trigger an error
+        # or a fallback if "fallback_to_host" is set in the config
+        return __non_queue
+
+    if is_torch_tensor(obj):
+        return get_torch_queue(obj)
+    else:
+        # no specialized queue can be extracted use or generate a generic. Note,
```
Comment (Vika-F) [inline onedal/utils/_sycl_queue_manager.py]:
```suggestion
        # no specialized queue can be extracted. Use or generate a generic. Note,
```
Or maybe I just do not understand the comment.

File: onedal/common/sycl.cpp [hunk 4]
```diff
@@ -42,7 +44,9 @@ void instantiate_sycl_interfaces(py::module& m) {
              [](const sycl::queue& queue) {
                  return pack_queue(std::make_shared<sycl::queue>(queue));
              })
-        .def_property_readonly("sycl_device", &sycl::queue::get_device);
+        .def_property_readonly("sycl_device", &sycl::queue::get_device)
+        .def(py::self == py::self)
```
Comment (david-cortes-intel) [inline onedal/common/sycl.cpp:48]: Wouldn't this need to compare pointer addresses of the queues instead?

File: sklearnex/tests/test_config.py [hunk 5]
```diff
@@ -205,3 +207,44 @@ def _onedal_test(self, *args, queue=None):
...
+    # by calling ``to_device``.
+    from onedal.utils import _sycl_queue_manager as QM
+    from sklearnex._device_offload import dispatch
+    from sklearnex._utils import PatchingConditionsChain
+
+    class FakeCUDA:
+        def __init__(self, data):
+            self.data = data
+
+        def to_device(self, *args):
```
Comment (david-cortes-intel) [inline sklearnex/tests/test_config.py:225]: Shouldn't it assert that this is called?

File: onedal/common/sycl.cpp (hunk 4 above)
Comment (icfaust) [reply onedal/common/sycl.cpp:48]: This exposes the underlying definition for ```sycl::queue::operator==``` which is most ideal. (https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/queue.hpp#L364  https://pybind11.readthedocs.io/en/stable/advanced/classes.html#operator-overloading

File: onedal/datatypes/_dlpack.py (hunk 1 above)
Comment (icfaust) [inline onedal/datatypes/_dlpack.py]:
```suggestion
        try:
            obj = np.from_dlpack(obj)
        except AttributeError:
            raise NotImplementedError("Upgrade NumPy >= 1.23 for dlpack support") from None 
```

File: sklearnex/tests/test_config.py [hunk 6]
```diff
@@ -205,3 +207,44 @@ def _onedal_test(self, *args, queue=None):
...
+            return self.data
```
Comment (icfaust) [inline sklearnex/tests/test_config.py]:
```suggestion
        def __init__(self, data):
            self.__data = data

        def to_device(self, *args):
            return self.__data
```

File: onedal/datatypes/_dlpack.py (hunk 1 above)
Comment (icfaust) [reply onedal/datatypes/_dlpack.py]: swapped to a try catch for minimal overhead.
